kv: Allow leases to be transfered back to self

andrewbaptist opened the pull request:
Previously a lease could be transferred using AdminTransferLease
to a different node, but a call to transfer back to the original
node was a no-op. This is done to prepare the work for the fix
to #73292. Specifically the fix there is to self-transfer the lease
to restart the lease timestamp and get correct protection on merges.

A few unit tests needed to be modified to make this work, as they
were assuming that AdminTransferLease was idempotent, when in
fact it was only lucky that they worked.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83157)
<!-- Reviewable:end -->

Commit:
kv: Allow leases to be transfered back to self

Previously a lease could be transferred using AdminTransferLease
to a different node, but a call to transfer back to the original
node was a no-op. This is done to prepare the work for the fix
to #73292. Specifically the fix there is to self-transfer the lease
to restart the lease timestamp and get correct protection on merges.

A few unit tests needed to be modified to make this work, as they
were assuming that AdminTransferLease was idempotent, when in
fact it was only lucky that they worked.

Release note: None
